tools/python/PythonTools.md

```diff
@@ -0,0 +1,165 @@
+    :param test_name: Name for test. Will be added to the root_path to create the test directory name.
+    :param name_input_map: Map of input names to numpy ndarray data for each input.
+    :param symbolic_dim_values_map: Map of symbolic dimension names to values to use for the input data if creating
+                                    using random data.
+    :param name_output_map: Optional map of output names to numpy ndarray expected output data.
+                            If not provided, the model will be run with the input to generate output data to save.
+    :return: None
+    """
+```
+
+Example usage:
+
+```python
+import sys
+import ort_test_dir_utils
+import onnx_test_data_utils
+
+# example model with two float32 inputs called 'input1' (dims: {2, 1}) and 'input2' (dims: {'dynamic', 4})
+model_path = '<onnxruntime root dir>/onnxruntime/test/testdata/transform/expand_elimination.onnx'
+
+# when using the default data generation any symbolic dimension values must be provided
+symbolic_vals = {'dynamic':2} # provide value for symbolic dim named 'dynamic' in 'input2'
+
+# let create_test_dir create random input in the (arbitrary) default range of -10 to 10. 
+# it will create data of the correct type based on the model.
+ort_test_dir_utils.create_test_dir(model_path, 'temp/examples', 'test1', symbolic_dim_values_map=symbolic_vals)
+
+# alternatively some or all input can be provided directly. any missing inputs will have random data generated.
+# symbolic dimension values are only required for input data that is randomly generated,
+# so we don't need to provide that in this case as we're explicitly providing all inputs.
+inputs = {'input1': np.array([100, 200]).reshape((2,1)).astype(np.float32),
+          'input2': np.random.randn(2,4).astype(np.float32)} # use 2 for the 'dynamic' dimension so shape is {2, 4}
+
+ort_test_dir_utils.create_test_dir(model_path, 'temp/examples', 'test2', name_input_map=inputs)
+
+# can easily dump the input and output to visually check it's as expected
+onnx_test_data_utils.dump_pb('temp/examples/test2/test_data_set_0')
+```
+
+### Running the test using python
+
+To execute the test once the directory is created you can use the onnx_test_runner or onnxruntime_perf_test executables if you have built onnxruntime from source, or the run_test_dir helper. Input can be either the test directory, or the model in case there are multiple in the test directory. 
+
+```python
+def run_test_dir(model_or_dir):
+    """
+    Run the test/s from a directory in ONNX test format.
+    All subdirectories with a prefix of 'test' are considered test input for one test run.
+
+    :param model_or_dir: Path to onnx model in test directory,
+                         or the test directory name if the directory only contains one .onnx model.
+    :return: None
+    """
+```
+
+Example usage:
+
+```python
+import sys
+import ort_test_dir_utils
+
+try:
+    ort_test_dir_utils.run_test_dir('temp/examples/test1')
```

Comment:
Some of the document instructions include non portable paths too `os.path.join('temp','examples','test1')`

Reply:
I believe Windows accepts either forward or backwards slashes.

```
D:\src\github\onnxruntime> python
Python 3.7.2 (tags/v3.7.2:9a3ffc0492, Dec 23 2018, 23:09:28) [MSC v.1916 64 bit (AMD64)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> import os
>>> os.chdir('/src/github')
>>> os.getcwd()
'D:\\src\\github'
```